README.md
# cr-xian

react component based on antV, for table data visualization

表格数据可视化为弦图


## ✨ Example View
<img src='./github/example2.png' alt='example' />

##  📦 Installation
```jsx
cnpm install --save cr-xian
or 
npm install --save cr-xian
```

## 🔨 Usage
<img src='./github/data-example.png' width='450px' alt='example' />

```js
import React, { PureComponent } from 'react';
import Xian from 'cr-xian';

const HHeader = ['张三', '李四', '小明']
const VHeader = ['华北', '华南']
const TData = [[20, 18, 42], [63, 51, 130]]

export default class App extends PureComponent {

  render () {
    return (
      <div>
        <Xian HHeader={HHeader} VHeader={VHeader} TData={TData}  />
      </div>
    )
  }
}
```

## 💡 Api
name | type | description
-----| -----| ------------
HHeader | array（必填） | 表头数组
VHeader | array（必填） | 列表头数组
TData | array（必填） | 表格主数据
width | number | 弦图宽度
height | number | 弦图高度
padding | array （例: [70, 50, 70, 50]）| 弦图内边距
outTitle | string | 弦外环 Tooltip 标题
innerTitle | string | 弦带 Tooltip 标题

##  Development

```
git clone https://github.com/Fea-Sin/cr-xian.git

cnpm install

npm start
```

## Example

http://localhost:8018/examples/testOne.html

## Test
```js
npm test

npm run coverage
```
## 🏁 Changelog
在 [HISTORY.md](./HISTORY.md) 中查看更新记录

## License

MIT


HISTORY.md
# History
----

## 1.0.2 / 2019-09-02
完善必填字段

## 1.0.0 / 2019-08-28

xian is start



tests/index.spec.js
import React from 'react'
import { mount } from 'enzyme'
import Comp from '../src'


describe('cr-comp', () => {

  it('base test comp', () => {
    // 跳过测试，偷懒！！！
    const num = 100
    expect(num).toBe(100)
    // const wrapper = mount(
    //   <Comp />
    // )
    // expect(wrapper.find('.hello').text()).toBe('弦图')
  })
})

src/chart.js
import React, { PureComponent } from 'react';
import DataSet from "@antv/data-set";
import { Chart, Tooltip, View, Coord, Geom, Label, Legend, Axis } from 'bizcharts';
import axios from 'axios';


class AppChart extends PureComponent {

  state = {
    dvForAll: [],
  }

  componentDidMount () {
    this.setData()
  }

  setData = () => {
    const ds = new DataSet({
      state: {
        currentState: '华南'
      }
    })
    axios.get('//39.107.35.212/csv/population-by-age.csv')
      .then(response => {
        const dvForAll = ds.createView('stateTest')
          .source(response.data, {
            type: 'csv',
          })

        dvForAll.transform({
          type: 'map',
          callback(row) {
            row['小明'] = Number(row['小明'])
            row['张三'] = Number(row['张三'])
            row['李四'] = Number(row['李四'])

            return row
          }
        })

        dvForAll.transform({
          type: 'percent',
          fields: '小明',
          dimension: 'state',
          as: 'percent-小明',
        })
        // dvForAll.transform({
        //   type: 'percent',
        //   fields: ['张三'],
        //   dimension: 'state',
        //   as: 'percent-张三',
        // })        
        // dvForAll.transform({
        //   type: 'percent',
        //   fields: ['李四'],
        //   dimension: 'state',
        //   as: 'percent-李四',
        // })
        dvForAll.transform({
          type: 'fold',
          fields: ['张三', '李四', '小明'],
          key: '姓名',
          value: '销量'
        })
        console.log('data set 处理后的数据----transform', dvForAll)

        this.setState({
          dvForAll,
        })
      })

  }

  render () {
    return (
      <div>
        <Chart
          data={this.state.dvForAll}
          height={400}        
        >
          <Legend />
          <Tooltip />
          <Axis name='姓名' />
          <Axis name="销量" />
          <Geom 
            type='intervalStack' 
            position='姓名*销量'
            color='state'
          />
        </Chart>
      </div>
    )
  }
}

export default AppChart;

src/index.js
import React, { PureComponent, Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import OuiDom from './utils/ouiDomUtils';
import AppChart from './chart';
import SChart from './chartT';
import { sum } from 'simple-statistics';

class App extends PureComponent {

  constructor(props) {
    super(props)
  }

  state = {
    chartData: {},
    loading: true,
  }

  componentDidMount() {
    this.getChartData()
    setTimeout(() => {
      this.setState({
        loading: false
      })
    }, 0)
  }
  componentDidUpdate(prevProps, prevState) {
  }
  
  getChartData = () => {
    const { HHeader, VHeader, TData } = this.props;
    const CNodes = []
    const CLinks = []
    let nodesArr = VHeader.concat(HHeader)
    let VSumArr = TData.map(item => {
      return sum(item)
    })

    let total = TData.reduce( (acc, cur) => {
      return acc + sum(cur)
    }, 0)

    let HSumArr = []
    for (let i=0; i<HHeader.length; i++) {
      let arr = []
       for (let j=0; j<VHeader.length; j++) {
         arr.push(TData[j][i])
       }
       HSumArr.push( sum(arr) )
    }
    let nodesSum = VSumArr.concat(HSumArr)

    // nodes data
    for (let i=0; i<nodesArr.length; i++) {
      let nodeObj = {}
      nodeObj.id = i
      nodeObj.name = nodesArr[i]
      nodeObj.sum = nodesSum[i]
      nodeObj.percent = ( (nodesSum[i] / total)*100 ).toFixed(2)
      CNodes.push(nodeObj)
    }

    // links data
    if (HHeader.length > 0) {
      for (let i=0; i<HHeader.length; i++) {
        let offset = VHeader.length
  
        for (let j=0; j<VHeader.length; j++) {
          let listObj = {}
          let listObjT = {}
          listObj.source = i+offset
          listObj.target = j  
          listObj.lineX = HHeader[i]
          listObj.lineY = VHeader[j]
          listObj.sourceWeight = TData[j][i]
          listObj.targetWeight = TData[j][i]    
  
          CLinks.push(listObj)
        }
      }
    }
    const chartData = {
      nodes: CNodes,
      links: CLinks,
    }
    // console.log('和弦图----', chartData)
    this.setState({
      chartData,
    })
  }

  render () {
    const { width, height, padding, outTitle, innerTitle } = this.props

    return (
      <div>
      {/* 数据可视化 */}
        {/* <AppChart /> */}
      {/* -弦图- */}
      {!this.state.loading && 
          <SChart 
            chartData={this.state.chartData}
            width={width}
            height={height}
            padding={padding}
            outTitle={outTitle}
            innerTitle={innerTitle}
          />
        }
      </div>
    )
  }
}

App.propTypes = {
  HHeader: PropTypes.array.isRequired,
  VHeader: PropTypes.array.isRequired,
  TData: PropTypes.arrayOf(PropTypes.array).isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  padding: PropTypes.arrayOf(PropTypes.number),
  outTitle: PropTypes.string,
  innerTitle: PropTypes.string,
}
App.defaultProps = {
  width: 550,
  height: 550,
  padding: [70, 50, 70, 50],
  outTitle: '地区销售合计百分比',
  innerTitle: '地区销售图',
}
export default App

src/chartT.js
import React, { PureComponent } from 'react';
import DataSet from "@antv/data-set";
import { Chart, Tooltip, View, Coord, Geom, Label } from 'bizcharts'

class SChart extends PureComponent {
  render () {
    const { chartData } = this.props
    const data = chartData
    const ds = new DataSet();
    const dv = ds.createView().source(data, {
      type: "graph",
      edges: d => d.links,
      nodes: d => d.nodes,
    });
    

    dv.transform({
      type: "diagram.arc",
      sourceWeight: edges => edges.sourceWeight,
      targetWeight: edges => edges.targetWeight,
      weight: true,
      marginRatio: 0.3,
      thickness: 0.1,
    });

    const scale = {
      x: {
        sync: true
      },
      y: {
        sync: true
      }
    };
    
    const { width, height, padding, outTitle, innerTitle } = this.props
    console.log('out---', outTitle, 'inner----', innerTitle)

    return (
      <div>
        <Chart
          data={data}
          height={height}
          width={width}
          scale={scale}
          padding={padding}
        >
          <Tooltip 
            showTitle={true} 
            inPlot={false}
          />
          <View data={dv.edges} axis={false}>
            <Coord type="polar" reflect="y" />
            <Geom
              type="edge"
              position="x*y"
              shape="arc"
              color="source"
              opacity={0.3}
              active={[true, { 
                highlight: true,
                style: {
                  fill: 'red'
                } 
              }]}
              tooltip={["lineX*lineY*sourceWeight", (lineX, lineY, sourceWeight) => {
                return {
                  title: `${outTitle}`,
                  name: `${lineY}-${lineX}：`,
                  value: `${sourceWeight}`
                }
              }]}
            />
          </View>
          <View data={dv.nodes} axis={false}>
            <Coord type="polar" reflect="y" />
            <Geom 
              type="polygon" 
              position="x*y" 
              color="id"
              tooltip={["name*sum*percent", (name, sum, percent) => {
                return {
                  title: `${innerTitle}`,
                  name: `${name}：`,
                  value: `合计: ${sum} 百分比: ${percent}`
                }
              }]}
              active={[true, { 
                highlight: true,
                style: {
                  fill: 'red'
                } 
              }]}

            >
              <Label
                content="name"
                labelEmit={true}
                textStyle={{
                  fill: "black"
                }}
              />
            </Geom>
          </View>
        </Chart>
      </div>
    )
  }
}

export default SChart

examples/testOne.js
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Testone from 'cr-xian';
import { Table } from 'antd';
import 'antd/dist/antd.css'; 
import '../assets/index.less';

const reactContainer = document.getElementById('__react-content');
const bodyContainer = document.getElementsByTagName('body')
bodyContainer[0].style.padding = '10px'
reactContainer.style.cssText = `
                                border: 1px solid #11d0bc;
                                padding: 10px;
                                margin-bottom: 10px;
                               `;

const dataSource = [
  {
    key: '1',
    address: '华北',
    zhangsan: '20',
    lisi: '18',
    xiaoming: '42',
  }, {
    key: '2',
    address: '华南',
    zhangsan: '63',
    lisi: '51',
    xiaoming: '130',
  }
];

const columns = [
  {
    title: '地区',
    dataIndex: 'address',
    key: 'address',
  }, {
    title: '张三',
    dataIndex: 'zhangsan',
    key: 'zhangsan',
  }, {
    title: '李四',
    dataIndex: 'lisi',
    key: 'lisi',
  }, {
    title: '小明',
    dataIndex: 'xiaoming',
    key: 'xiaoming',
  }
]

const HHeader = ['张三', '李四', '小明']
const VHeader = ['华北', '华南']
const TData = [[20, 18, 42], [63, 51, 130]]

function render(container) {
  ReactDOM.render(
    <div>
      <h2>弦图</h2>
      <div style={{lineHeight: '40px'}}>表格可视化为弦图</div>
      <Table dataSource={dataSource} columns={columns} pagination={false} />
      <div style={{lineHeight: '40px'}}>将表格转化为弦图</div>
      <div style={{width: 800, height: 800, margin: '0 auto'}}>
        <Testone 
          HHeader={HHeader} VHeader={VHeader} TData={TData}
          width={500}
          height={500}
        />
      </div>

    </div>, container
  )
}

render(reactContainer)